### PR TITLE
Towards knots

### PR DESCRIPTION
**Pull request by fieker**

### CODE REVIEW

File: src/GrpAb/GrpAbFinGen.jl [hunk 1]
```diff
@@ -1872,12 +1872,133 @@ end
...
+julia> U, _ = sub(G, [G[1]+4*G[2]]);
+
+julia> is_pure(U)
+true
+
+julia> has_complement(U, G)[1]
+true
+```
+"""
+function is_pure(U::GrpAbFinGen, G::GrpAbFinGen)
```
Comment (thofma) [inline src/GrpAb/GrpAbFinGen.jl:1905]: Is this different from `G/U` being torsion-free?

File: src/GrpAb/GrpAbFinGen.jl [hunk 2]
```diff
@@ -1872,12 +1872,133 @@ end
...
+  @assert is_finite(G)
+  #not implemented in general: needs to be done via torsion subgroup
+  n = exponent(G)
+  lf = factor(n).fac
```
Comment (thofma) [inline src/GrpAb/GrpAbFinGen.jl:1909]:
```suggestion
  lf = factor(n)
```

File: src/GrpAb/GrpAbFinGen.jl [hunk 3]
```diff
@@ -1872,12 +1872,133 @@ end
...
+      end
+    end
+  end
+  return true
+end
+
+@doc Markdown.doc"""
+    is_neat(U::GrpAbFinGen, G::GrpAbFinGen) -> Bool
+
+A subgroup `U` of `G` is called pure iff for all primes `p` an element in `U`
```
Comment (thofma) [inline src/GrpAb/GrpAbFinGen.jl:1925]:
```suggestion
A subgroup `U` of `G` is called neat iff for all primes `p` an element in `U` 
```
?

**Comment by fieker** [comment]
On Tue, Dec 13, 2022 at 06:08:49AM -0800, Tommy Hofmann wrote:
> @thofma commented on this pull request.
> 
> 
> 
> > +
> +julia> U, _ = sub(G, [G[1]+2*G[2]]);
> +
> +julia> is_pure(U, G)
> +false
> +
> +julia> U, _ = sub(G, [G[1]+4*G[2]]);
> +
> +julia> is_pure(U)
> +true
> +
> +julia> has_complement(U, G)[1]
> +true
> +```
> +"""
> +function is_pure(U::GrpAbFinGen, G::GrpAbFinGen)
> 
> Is this different from `G/U` being torsion-free?
Yep. Im my context G and U are finite, hence nothing is torsion free

**Comment by fieker** [comment]
On Tue, Dec 13, 2022 at 06:09:15AM -0800, Tommy Hofmann wrote:
> @thofma commented on this pull request.
> 
> 
> 
> > +false
> +
> +julia> U, _ = sub(G, [G[1]+4*G[2]]);
> +
> +julia> is_pure(U)
> +true
> +
> +julia> has_complement(U, G)[1]
> +true
> +```
> +"""
> +function is_pure(U::GrpAbFinGen, G::GrpAbFinGen)
> +  @assert is_finite(G)
> +  #not implemented in general: needs to be done via torsion subgroup
> +  n = exponent(G)
> +  lf = factor(n).fac
> 
> ```suggestion
>   lf = factor(n)
> ```
I am never sure what interface factor allows... (p,k) yes, but
keys(lf.fac) has no nice interface?

**Comment by fieker** [comment]
On Tue, Dec 13, 2022 at 06:09:36AM -0800, Tommy Hofmann wrote:
> @thofma commented on this pull request.
> 
> 
> 
> > +  for (p, k) = lf
> +    for i=1:k
> +      h = hom(G, G, [p^i*g for g = gens(G)])
> +      q, mq = quo(intersect(image(h)[1], U), h(U)[1])
> +      if order(q) != 1
> +        return false
> +      end
> +    end
> +  end
> +  return true
> +end
> +
> ***@***.*** Markdown.doc"""
> +    is_neat(U::GrpAbFinGen, G::GrpAbFinGen) -> Bool
> +
> +A subgroup `U` of `G` is called pure iff for all primes `p` an element in `U` 
> 
> ```suggestion
> A subgroup `U` of `G` is called neat iff for all primes `p` an element in `U` 
> ```
> ?
definitely

**Comment by thofma** [comment]
> Yep. Im my context G and U are finite, hence nothing is torsion free

Ah, yes, that makes sense.

Regarding the factorization. You are right, there is no interface to `keys(lf.fac)`.

**Pull request merged**